### PR TITLE
implement nt path conversion for windows

### LINKED ISSUE #6044: '.' and '..' components in relative paths in std.os.windows.OpenFile
While working on #5701 and #5993 I've noticed that doing something like this on Windows:

```zig
const fd: fd_t = some_fd_thats_non_null;
const pathname = [_]u16{ '.' }; // or '..', or in fact, having '.' or '..' anywhere in the path
const new_fd = try std.os.windows.OpenFile(pathname, .{ .dir = fd });
defer std.os.windows.CloseHandle(new_fd);
```

will yield a runtime panic (actually, `NtCreateFile` will trip on `OBJECT_NAME_INVALID`). As far as I can work out, this is due to the presence of '.' and '..' in the pathname argument. My question is here, is there any simple way around this, or are we forced to re-implement `CreateFileW` where we either query the OS for the handle to cwd, or we use the `.dir` handle, get the canonical path to it using `GetFinalPathNameByHandle`, concatenate the resultant path with the specified argument path, and feed this path into `NtCreateFile` again (not sure if that would even work since I'd imagine we'd prepend `\??\` to the resultant path which could potentially not accept '.' and '..' as valid path components)?

I'm not yet convinced whether we'd want to enable this kind of behaviour for all calls using `std.os.windows.OpenFile`, however, it does seem useful for `std.os.realpath` and `std.fs.Dir.realpath`, when canonicalizing the pathnames.

### PR DESCRIPTION
**Pull request by marler8997**
Second attempt to fix #6044  (previous attempt https://github.com/ziglang/zig/pull/7537)

The first attempt leveraged the functionality in `ntdll` to perform the Dos to NT path conversion, however, that API requires heap allocation which @andrewrk has indicated we don't want (see https://github.com/ziglang/zig/pull/7537#discussion_r550807530).

So this PR implements the logic to convert a DOS file path into an NT-compatible path in Zig's std lib.

### CODE REVIEW

File: lib/std/os/windows.zig [hunk 1]
```diff
@@ -1528,6 +1528,81 @@ pub const PathSpace = struct {
     }
 };
 
+/// The error type for `removeDotDirsSanitized`
+pub const RemoveDotDirsError = error { TooManyParentDirs };
+
+/// Removes '.' and '..' path components from a "sanitized relative path".
```
Comment (daurnimator) [inline lib/std/os/windows.zig:1534]: this function looks like it belongs in std.fs.path. it might be a duple of `resolveWindows`?

Comment (marler8997) [reply lib/std/os/windows.zig:1534]: It's specific to windows...only works with backslashes.  We could also make it work with forward slashes but I think posix has other better way to perform this operation.  This function doesn't touch the filesystem or use `CWD` to normalize the path like `resolveWindows` does.

Comment (daurnimator) [reply lib/std/os/windows.zig:1534]: > It's specific to windows...only works with backslashes

Os-specific code is all over the standard library: it does all just get dumped into os/windows.zig.

File: lib/std/mem.zig [hunk 2]
```diff
@@ -1829,6 +1829,49 @@ test "replace" {
     testing.expect(eql(u8, output[0..], "Favor reading  over writing ."));
 }
 
+/// Replace all occurences of `needle` with `replacement`.
+pub fn replaceElem(comptime T: type, slice: []T, needle: T, replacement: T) void {
```
Comment (LemonBoy) [inline lib/std/mem.zig:1833]: If we follow the `std.mem` naming this should be `replaceScalar` and take an `output: []T` parameter (slice becomes `const`).
Specifying the same in/out buffers gets you the in-place replacement you want.

File: lib/std/mem.zig [hunk 3]
```diff
@@ -1829,6 +1829,49 @@ test "replace" {
...
+    for (slice) |e, i| {
+        if (e == needle) {
+            slice[i] = replacement;
+        }
+    }
+}
+
+/// Collapse consecutive duplicate elements into one entry.
+pub fn collapseRepeats(comptime T: type, slice: []T, needle: T) usize {
```
Comment (LemonBoy) [inline lib/std/mem.zig:1842]: Not much of a needle, it's more a `elem`ent no?
And given it squishes `slice` contents it makes sense for this to return a slice of it.

File: lib/std/mem.zig [hunk 4]
```diff
@@ -1829,6 +1829,49 @@ test "replace" {
...
+    if (slice.len == 0) return 0;
+    var write_idx : usize = 1;
```
Comment (LemonBoy) [inline lib/std/mem.zig:1844]: Smells like unformatted code...

Comment (marler8997) [reply lib/std/mem.zig]: yeah, should be formatted now

File: lib/std/mem.zig (hunk 2 above)
Comment (marler8997) [reply lib/std/mem.zig]: fixed

File: lib/std/mem.zig (hunk 3 above)
Comment (marler8997) [reply lib/std/mem.zig]: I've changed `needle` to `elem`.

As for returning a slice, such an interface could be supported with a wrapper function at no runtime cost. In the current uses, only the resulting size is used and this same interface is propagated in a handful of places which makes the code simpler.

Comment (LemonBoy) [reply lib/std/mem.zig]: You can get the old behaviour by simply adding a `.len` at the end.
Returning a slice makes the API much easier to use as `f(collapseRepeats(..))`, throwing indices around is error-prone.

Comment (marler8997) [reply lib/std/mem.zig]: > Returning a slice makes the API much easier to use

In some cases yes.  If such a use case comes up we can add a wrapper function that does this.

Comment (LemonBoy) [reply lib/std/mem.zig]: It should be the other way around, returning a slice returns a pointer _and_ the adjusted length.

Comment (marler8997) [reply lib/std/mem.zig]: > It should be the other way around, returning a slice returns a pointer and the adjusted length.

My philosphy is to do less, and then only do more if you need to.  Going the other way around means always doing more even if you don't need to, and then undoing what you don't need.

Comment (marler8997) [reply lib/std/mem.zig]: Another example of an API that uses this methodology is the Allocator `resizeFn` function.

Comment (LemonBoy) [reply lib/std/mem.zig]: If you're adding new public-facing APIs you can't just think in terms of "do less" or think of your single use-case.
Returning the length makes using this function with sliced inputs hard to use, consider the following example:
```zig
const x = collapseRepeats(foo[bar..Baz], ...)
```
If you return a sub-slice the user is able to quickly use the modified chunk of input, while if the length is returned instead another slicing operation is required.

But given that opinions are like assholes I'll rest my case, I'm getting too old for this shit.

Comment (marler8997) [reply lib/std/mem.zig]: Ok the fact that this is a public facing function is a good point.  How about creating a `collapseRepeatsLen` and `collapseRepeats` that wraps it?

File: lib/std/fs.zig [hunk 5]
```diff
@@ -1285,15 +1285,6 @@ pub const Dir = struct {
             .SecurityDescriptor = null,
             .SecurityQualityOfService = null,
         };
-        if (sub_path_w[0] == '.' and sub_path_w[1] == 0) {
-            // Windows does not recognize this, but it does work with empty string.
-            nt_name.Length = 0;
-        }
-        if (sub_path_w[0] == '.' and sub_path_w[1] == '.' and sub_path_w[2] == 0) {
-            // If you're looking to contribute to zig and fix this, see here for an example of how to
-            // implement this: https://git.midipix.org/ntapi/tree/src/fs/ntapi_tt_open_physical_parent_directory.c
```
Comment (andrewrk) [inline lib/std/fs.zig:1294]: I don't see this addressed in the PR description? Did you look into using this technique to resolve `..` in a way that respects following symlinks?

**Review by andrewrk — CHANGES_REQUESTED** [review]
Thanks for the patch. This solution involves manipulating path strings, however Windows has symlinks, and on other platforms, when opening files with `std.fs`, ".." path components respect symlinks, so we need to do it here, too.

File: lib/std/fs.zig (hunk 5 above)
Comment (marler8997) [reply lib/std/fs.zig:1294]: On windows symlinks are resolved after resolving ".." (https://github.com/ziglang/zig/issues/7751)

**Comment by squeek502** [comment]
Nice work @marler8997! This hugely improves `std.fs` on Windows.

**Comment by andrewrk** [comment]
Thanks for the clean-up PRs @squeek502!

**Pull request merged**